From 101daf566661f208f78081de2d7d4040141ea5e1 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Tue, 6 Oct 2020 15:03:13 -0700
Subject: [PATCH 1/4] co-located: add a config option to indicate co locate

A 6GHz BSS can be co-located with a 2/5GHz BSS. This option allows us to
indicate if this should be done.

Signed-off-by: John Crispin <john@phrozen.org>
---
 hostapd/config_file.c | 2 ++
 hostapd/hostapd.conf  | 3 +++
 src/ap/ap_config.h    | 1 +
 3 files changed, 6 insertions(+)

--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -3680,6 +3680,8 @@ static int hostapd_config_fill(struct ho
 			return 1;
 		}
 		bss->unsol_bcast_probe_resp_interval = val;
+	} else if (os_strcmp(buf, "he_co_locate") == 0) {
+		conf->he_co_locate = atoi(pos);
 #endif /* CONFIG_IEEE80211AX */
 	} else if (os_strcmp(buf, "max_listen_interval") == 0) {
 		bss->max_listen_interval = atoi(pos);
--- a/hostapd/hostapd.conf
+++ b/hostapd/hostapd.conf
@@ -581,6 +581,9 @@ wmm_ac_vo_acm=0
 # Default: 0 (disabled)
 #notify_mgmt_frames=0
 
+# Enable co-locate for a 6GHz radio
+#co_locate=0
+
 ##### IEEE 802.11n related configuration ######################################
 
 # ieee80211n: Whether IEEE 802.11n (HT) is enabled
-- 
2.7.4

From e9263d5c0a8e166e025a63f823288266d1085942 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Tue, 6 Oct 2020 15:05:24 -0700
Subject: [PATCH 2/4] neighbor_db: set the co-locate bit

If the BSS is co-located, the corresponding bit needs to be set inside
bssid_info.

Signed-off-by: John Crispin <john@phrozen.org>
---
 src/ap/neighbor_db.c         | 3 +++
 src/common/ieee802_11_defs.h | 2 ++
 2 files changed, 5 insertions(+)


--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1086,6 +1086,7 @@ struct hostapd_config {
 	u8 he_6ghz_max_ampdu_len_exp;
 	u8 he_6ghz_rx_ant_pat;
 	u8 he_6ghz_tx_ant_pat;
+	bool he_co_locate;
 #endif /* CONFIG_IEEE80211AX */
 
 	/* VHT enable/disable config from CHAN_SWITCH */
-- 
2.7.4

From 92d786ddb2568b688ed4d25e32f3987e926c0ca4 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Mon, 21 Sep 2020 13:50:58 +0200
Subject: [PATCH 3/4] rrm: add handling for co-located BSS

A BSS shall include the neighbor entry of a co-located BSS upon an incoming
request.

Signed-off-by: John Crispin <john@phrozen.org>
---
 src/ap/rrm.c | 117 +++++++++++++++++++++++++++++++++++++----------------------
 1 file changed, 74 insertions(+), 43 deletions(-)


--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7236,7 +7236,7 @@ static size_t hostapd_eid_rnr_colocation
 		iface = hapd->iface->interfaces->iface[i];
 
 		if (iface == hapd->iface ||
-		    !is_6ghz_op_class(iface->conf->op_class))
+		    !is_6ghz_op_class(iface->conf->op_class) || !iface->conf->he_co_locate)
 			continue;
 
 		len += hostapd_eid_rnr_iface_len(iface->bss[0], hapd,
@@ -7402,7 +7402,8 @@ static u8 * hostapd_eid_rnr_iface(struct
 				bss_param |=
 					RNR_BSS_PARAM_UNSOLIC_PROBE_RESP_ACTIVE;
 
-			bss_param |= RNR_BSS_PARAM_CO_LOCATED;
+			if (bss->iconf->he_co_locate)
+				bss_param |= RNR_BSS_PARAM_CO_LOCATED;
 
 			*eid++ = bss_param;
 			*eid++ = RNR_20_MHZ_PSD_MAX_TXPOWER - 1;
@@ -7436,7 +7437,7 @@ static u8 * hostapd_eid_rnr_colocation(s
 		iface = hapd->iface->interfaces->iface[i];
 
 		if (iface == hapd->iface ||
-		    !is_6ghz_op_class(iface->conf->op_class))
+		    !is_6ghz_op_class(iface->conf->op_class) || !iface->conf->he_co_locate)
 			continue;
 
 		eid = hostapd_eid_rnr_iface(iface->bss[0], hapd, eid,
--- a/src/ap/neighbor_db.c
+++ b/src/ap/neighbor_db.c
@@ -282,6 +282,9 @@ void hostapd_neighbor_set_own_report(str
 			bssid_info |= NEI_REP_BSSID_INFO_HE;
 	}
 
+	if (hapd->iconf->he_co_locate)
+		bssid_info |= NEI_REP_BSSID_INFO_HE_CO_LOCATED;
+
 	/* TODO: Set NEI_REP_BSSID_INFO_MOBILITY_DOMAIN if MDE is set */
 
 	if (ieee80211_freq_to_channel_ext(hapd->iface->freq,
-- 
2.7.4

From fe0833e906559efb41ec9ee3f3e77049b968e6aa Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Wed, 21 Oct 2020 11:47:22 -0700
Subject: [PATCH 4/4] rnr: add reduced neighbor reporting

The Reduced Neighbor Report (rnr) element contains channel and other
information related to neighbor APs. It is part of the OCE requirement.

Signed-off-by: John Crispin <john@phrozen.org>
Signed-off-by: Aloka Dixit <alokad@codeaurora.org>
Signed-off-by: Muna Sinada <msinada@codeaurora.org>
---
 src/ap/beacon.c              |  40 +++++++++-
 src/ap/ieee802_11.c          | 177 ++++++++++++++++++++++++++++++++-----------
 src/ap/ieee802_11.h          |   4 +-
 src/common/ieee802_11_defs.h |   6 +-
 4 files changed, 173 insertions(+), 54 deletions(-)


--- a/src/common/ieee802_11_defs.h
+++ b/src/common/ieee802_11_defs.h
@@ -2169,6 +2169,8 @@ enum phy_type {
 #define NEI_REP_BSSID_INFO_VHT BIT(12)
 #define NEI_REP_BSSID_INFO_FTM BIT(13)
 #define NEI_REP_BSSID_INFO_HE BIT(14)
+#define NEI_REP_BSSID_INFO_HE_ER_BSS BIT(15)
+#define NEI_REP_BSSID_INFO_HE_CO_LOCATED BIT(16)
 
 /*
  * IEEE P802.11-REVmc/D5.0 Table 9-152 - HT/VHT Operation Information
